File: packages/plugin/ios/Plugin/NativeNavigationPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(NativeNavigationPlugin)
public class NativeNavigationPlugin: CAPPlugin {
    private var implementation: NativeNavigation! = nil

    @objc override public func load() {
        self.implementation = NativeNavigation(bridge: self.bridge!, plugin: self)
    }

    @objc func present(_ call: CAPPluginCall) {
        do {
            let options = try PresentOptions.fromJSObject(call)

            Task {
                do {
                    let result = try await implementation.present(options)
                    call.resolve(result.toPluginResult())
                } catch {
                    call.reject("Failed to present: \(error.localizedDescription)")
                }
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func dismiss(_ call: CAPPluginCall) {
        do {
            let options = try DismissOptions.fromJSObject(call)
            
            Task {
                do {
                    let result = try await implementation.dismiss(options)
                    call.resolve(result.toPluginResult())
                } catch {
                    call.reject("Failed to dismiss: \(error.localizedDescription)")
                }
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func push(_ call: CAPPluginCall) {
        do {
            let options = try PushOptions.fromJSObject(call)

            Task {
                do {
                    let result = try await implementation.push(options)
                    call.resolve(result.toPluginResult())
                } catch {
                    call.reject("Failed to push: \(error.localizedDescription)")
                }
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func pop(_ call: CAPPluginCall) {
        do {
            let options = try PopOptions.fromJSObject(call)
            
            Task {
                do {
                    let result = try await implementation.pop(options)
                    call.resolve(result.toPluginResult())
                } catch {
                    call.reject("Failed to pop: \(error.localizedDescription)")
                }
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func update(_ call: CAPPluginCall) {
        do {
            guard let id = call.getString("id") else {
                throw NativeNavigatorError.missingParameter(name: "id")
            }
            
            guard let implementation = self.implementation else {
                throw NativeNavigatorError.illegalState(message: "Implementation is missing")
            }
            
            let updateOptions = try UpdateOptions.fromJSObject(call)
            if let options = updateOptions.update {
                var componentSpec = try implementation.findComponent(id: id).spec
                try componentSpec.update(options)
                let updatedSpec = componentSpec
                do {
                    try implementation.update(updateOptions, updatedSpec: updatedSpec)
                    call.resolve()
                } catch {
                    call.reject("Failed to update: \(error.localizedDescription)")
                }
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc public func reset(_ call: CAPPluginCall) {
        do {
            let options = try ResetOptions.fromJSObject(call)
            
            Task {
                do {
                    try await implementation.reset(options)
                    call.resolve()
                } catch {
                    call.reject("Failed to reset: \(error.localizedDescription)")
                }
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func get(_ call: CAPPluginCall) {
        do {
            let options = try GetOptions.fromJSObject(call)
            Task {
                do {
                    let result = try await implementation.get(options)
                    call.resolve(result.toPluginResult())
                } catch {
                    call.reject("Failed to get: \(error.localizedDescription)")
                }
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc func message(_ call: CAPPluginCall) {
        do {
            let options = try MessageOptions.fromJSObject(call)
            Task {
                do {
                    try await implementation.message(options)
                    call.resolve()
                } catch {
                    call.reject("Failed to message: \(error.localizedDescription)")
                }
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }
    
    @objc public func viewReady(_ call: CAPPluginCall) {
        do {
            let options = try ViewReadyOptions.fromJSObject(call)
            
            Task {
                do {
                    try await implementation.viewReady(options)
                    call.resolve()
                } catch {
                    call.reject("Failed to handle view ready: \(error.localizedDescription)")
                }
            }
        } catch {
            call.reject(error.localizedDescription)
        }
    }

}

File: packages/plugin/android/src/main/java/com/cactuslab/capacitor/nativenavigation/NativeNavigationPlugin.kt
package com.cactuslab.capacitor.nativenavigation

import android.net.Uri
import android.util.Log
import android.webkit.WebChromeClient
import android.webkit.WebView
import androidx.lifecycle.ViewModelProvider
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import com.cactuslab.capacitor.nativenavigation.exceptions.MissingParameterException
import com.cactuslab.capacitor.nativenavigation.types.*
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin

@CapacitorPlugin(name = "NativeNavigation")
class NativeNavigationPlugin : Plugin() {
    private lateinit var implementation: NativeNavigation
    private lateinit var model: NativeNavigationViewModel

    private var isLoaded = false

    private val webViewListener = object: WebViewListener() {
        override fun onPageStarted(webView: WebView?) {
            Log.d(TAG, "onPageStarted: A page start was detected on ${webView}")
            if (webView == bridge.webView) {
                Log.d(TAG, "onPageStarted: A page start was detected on the root view. Calling reset now.")
                activity.runOnUiThread {
                    cleanUp()
                }
            }
        }
    }

    override fun load() {
        model = ViewModelProvider(activity)[NativeNavigationViewModel::class.java]
        implementation = NativeNavigation(this, model)
        isLoaded = true

    }

    override fun handleOnStart() {
        super.handleOnStart()
        bridge.addWebViewListener(webViewListener)
    }

    override fun handleOnDestroy() {
        super.handleOnDestroy()
        bridge.removeWebViewListener(webViewListener)
    }

    override fun shouldOverrideLoad(url: Uri?): Boolean? {
        if (!isLoaded) {
            return null
        }
        val result = implementation.shouldOverrideLoad(url)
        Log.d(TAG, "ShouldOverrideLoad of url $url returning $result")
        return result
    }

    @PluginMethod
    fun viewReady(call: PluginCall) {
        try {
            val options = ViewReadyOptions.fromJSObject(call.data)
            implementation.viewReady(options = options)
            call.resolve()
        } catch (e: MissingParameterException) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun present(call: PluginCall) {
        activity.runOnUiThread {
            try {
                val chromeClient = capacitorChromeClient()
                bridge.webView.webChromeClient = NativeNavigationChromeClient(chromeClient, implementation)
                bridge.webView.settings.setSupportMultipleWindows(true)

                val options = PresentOptions.fromJSObject(call.data)
                implementation.present(options = options, call = call)
            } catch (e: MissingParameterException) {
                call.reject(e.localizedMessage)
            }
        }
    }

    @PluginMethod
    fun dismiss(call: PluginCall) {
        try {
            val options = DismissOptions.fromJSObject(call.data)
            activity.runOnUiThread {
                implementation.dismiss(options, call)
            }
        } catch (e: MissingParameterException) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun push(call: PluginCall) {
        try {
            val options = PushOptions.fromJSObject(call.data)
            activity.runOnUiThread {
                implementation.push(options = options, call = call)
            }
        } catch (e: MissingParameterException) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun pop(call: PluginCall) {

        activity.runOnUiThread {
            implementation.pop(call, activity)
        }
//        call.reject("Pop not ready")
//        call.resolve()
    }

    @PluginMethod
    fun get(call: PluginCall) {
        try {
            val options = GetOptions.fromJSObject(call.data)
            activity.runOnUiThread {
                implementation.getOptions(options, call)
            }
        } catch (e: MissingParameterException) {
            call.reject(e.localizedMessage)
        }


    }

    @PluginMethod
    fun update(call: PluginCall) {
        try {
            val options = UpdateOptions.fromJSObject(call.data)
            activity.runOnUiThread {
                implementation.update(options)
                call.resolve()
            }
        } catch (e: MissingParameterException) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    fun message(call: PluginCall) {
        try {
            val options = MessageOptions.fromJSObject(call.data)
            activity.runOnUiThread {
                implementation.message(options, call)
            }
        } catch (e: MissingParameterException) {
            call.reject(e.localizedMessage)
        }
    }

    private fun capacitorChromeClient(): WebChromeClient = if (WebViewFeature.isFeatureSupported(WebViewFeature.GET_WEB_CHROME_CLIENT)) {
        when (val client = WebViewCompat.getWebChromeClient(bridge.webView)) {
            is BridgeWebChromeClient -> {
                client
            }
            is NativeNavigationChromeClient -> {
                client.bridgeChromeClient
            }
            else -> client ?: throw Exception("Unexpected web client")
        }
    } else {
        BridgeWebChromeClient(bridge)
    }


    private fun cleanUp() {
        val client = capacitorChromeClient()
        bridge.webView.webChromeClient = client

        implementation.reset()
    }

    @PluginMethod
    fun reset(call: PluginCall) {
        // This is a tear down method. Reset the UI to the capacitor state.
        activity.runOnUiThread {
            cleanUp()
            call.resolve()
        }
    }

    fun notifyCreateView(path: String?, id: String, state: JSObject?, stack: String?) {
        val obj = JSObject()
        path?.let { obj.put("path", it) }
        obj.put("id", id)
        state?.let {
            obj.put("state", it)
        }
        stack?.let {
            obj.put("stack", stack)
        }

        Log.d(TAG, "Notify Create View [path: $path, id: $id, state: $state]")

        notifyListeners("createView", obj, true)
    }

    fun notifyUpdateView(path: String?, id: String, state: JSObject?, stack: String?) {
        val obj = JSObject()
        path?.let { obj.put("path", it) }
        obj.put("id", id)
        state?.let {
            obj.put("state", it)
        }
        stack?.let {
            obj.put("stack", stack)
        }

        Log.d(TAG, "Notify Update View [path: $path, id: $id, state: $state]")

        notifyListeners("updateView", obj, true)
    }

    fun notifyDestroyView(id: String) {
        val obj = JSObject()
        obj.put("id", id)
        Log.d(TAG, "Notify Destroy View [id: $id]")

        notifyListeners("destroyView", obj, true)
    }

    fun notifyClick(buttonId: String, componentId: String) {
        val obj = JSObject()
        obj.put("buttonId", buttonId)
        obj.put("componentId", componentId)

        notifyListeners("click:$componentId",  obj, true)
        notifyListeners("click", obj, true)
    }

    fun notifyMessage(target: String, type: String, value: JSObject?) {
        val obj = JSObject()
        obj.put("target", target)
        obj.put("type", type)
        value?.let {
            obj.put("value", it)
        }

        notifyListeners("message", obj, true)
    }

    fun notifyViewWillAppear(componentId: String) {
        notifyListeners("viewWillAppear:$componentId", JSObject(), true)
    }

    fun notifyViewDidAppear(componentId: String) {
        notifyListeners("viewDidAppear:$componentId", JSObject(), true)
    }

    fun notifyViewWillDisappear(componentId: String) {
        notifyListeners("viewWillDisappear:$componentId", JSObject(), true)
    }

    fun notifyViewDidDisappear(componentId: String) {
        notifyListeners("viewDidDisappear:$componentId", JSObject(), true)
    }

    companion object {
        const val TAG = "NNPlugin"
    }

}